### PR TITLE
feat: Autonomous mode

### PR DESCRIPTION
**Pull request by SmartManoj**
Solves #1798.

image

For Future PRs: continue the previous task

### CODE REVIEW

File: agenthub/codeact_agent/codeact_agent.py
Comment (yufansong) [inline agenthub/codeact_agent/codeact_agent.py:228]: Why add this

File: agenthub/codeact_agent/action_parser.py
Comment (yufansong) [inline agenthub/codeact_agent/action_parser.py:183]: QQ: Will the config value change when running?

**Review by yufansong — CHANGES_REQUESTED** [review]
cc @li-boxuan  @xingyaoww  for review.

File: agenthub/codeact_agent/action_parser.py
Comment (SmartManoj) [reply agenthub/codeact_agent/action_parser.py:183]: Not in this PR

File: agenthub/codeact_agent/codeact_agent.py
Comment (SmartManoj) [reply agenthub/codeact_agent/codeact_agent.py:228]: Without this, the agent repeatedly gives step 1 for the given task.

Comment (SmartManoj) [reply agenthub/codeact_agent/codeact_agent.py:228]: Also, it gives a hint about how many steps it has tried.

File: agenthub/codeact_agent/action_parser.py
Comment (li-boxuan) [inline agenthub/codeact_agent/action_parser.py:191]: this is too hacky. This class parses the response from LLM, and we shouldn't append response here.

File: agenthub/codeact_agent/codeact_agent.py
Comment (li-boxuan) [reply agenthub/codeact_agent/codeact_agent.py:228]: I don't understand it either. Doesn't look right to me...

Comment (xingyaoww) [reply agenthub/codeact_agent/codeact_agent.py:228]: I don't think `role == 'agent'` is a supported behavior of litellm, it should be `user` message? Did you test the code?

**Review by xingyaoww — CHANGES_REQUESTED** [review]
I'm a bit hestitant to merge this as is, unless we figure out a way to add a checkbox on the frontend that allows the user to select [autonomous mode]. 

I think a lot of OpenDevin's helpfulness comes from the ability to request inputs from the user, and i think it is probably not a good idea to completely remove this ability just to implement autonomous mode.

**Comment by tobitege** [comment]
> I'm a bit hestitant to merge this as is, unless we figure out a way to add a checkbox on the frontend that allows the user to select [autonomous mode].
> 
> I think a lot of OpenDevin's helpfulness comes from the ability to request inputs from the user, and i think it is probably not a good idea to completely remove this ability just to implement autonomous mode.

I'd rather have an instantly working `Stop` frontend button that can interrup even a running task, without waiting for the current one to finish. Let the user control. This PR's behavior seems highly LLM dependent.

**Comment by xingyaoww** [comment]
> I'd rather have an instantly working Stop frontend button that can interrupt even a running task, without waiting for the current one to finish. Let the user control. 

OHHH yes! This is what i want too -- something like `ctrl+C`

**Comment by tobitege** [comment]
Is this PR just for a specific edge-case and does it warrant _patching around the core_ that make it less maintainable?

**Comment by SmartManoj** [comment]
> i think it is probably not a good idea to completely remove this ability

@xingyaoww, it is still a config.

**Comment by xingyaoww** [comment]
> it is still a config.

1. If you don't teach the user to tweak the config, there's just no effect from a user's perspective -- we are writing code that's not actually useful to most users since most of them will start OpenDevin using docker (and by default, without autonomous mode).

2. If the user configures the autonomous mode -- We throw away the ability to await user input, which is one of our core features. Moreover, we currently rely on some hacks to do it, which creates a future maintenance burden.

Therefore, I don't think it is good to merge unless we can solve (1) in a meaningful way - bc if (1) is not solved, user experience is not gonna change, and it basically means we are adding code that is not actively used by people.

**Comment by SmartManoj** [comment]
It's currently for advances users and completely solves #1586

File: agenthub/codeact_agent/codeact_agent.py
Comment (SmartManoj) [reply agenthub/codeact_agent/codeact_agent.py:228]: Litellm doesn't check that but openai does. Ran with a custom model.

**Comment by mamoodi** [comment]
@SmartManoj based on the comments by @xingyaoww seems like this might not be feasible as is. Will you be continuing to address the comments or should this be closed?

**Comment by neubig** [comment]
Because there has been no response to @mamoodi 's comments above, I am going to close this PR, but please feel free to explain and reopen if you think otherwise.

**Comment by SmartManoj** [comment]
> based on the comments by @xingyaoww seems like this might not be feasible as is.

There are some commits and a request review after that.

image

**Review by xingyaoww — COMMENTED** [review]
Sorry it takes so long for review --Just saw it!

I think this version looks good to me, just some nits, and I'll need to locally test this before approve it.

@SmartManoj can you share a screenshot of the modified frontend so reviewer can see its effect?

**Comment by SmartManoj** [comment]
> can you share a screenshot of the modified frontend so reviewer can see its effect?

Added in the description.

File: frontend/src/components/chat/ChatInterface.tsx
Comment (xingyaoww) [inline frontend/src/components/chat/ChatInterface.tsx:136]:
```suggestion
            Autonomous Mode
```

Comment (SmartManoj) [reply frontend/src/components/chat/ChatInterface.tsx]: or Auto mode?

Comment (xingyaoww) [reply frontend/src/components/chat/ChatInterface.tsx]: Sounds good!

**Comment by xingyaoww** [comment]
The tooltip is a bit laggy in this PR.
I suggested a change by re-using the `Tooltip` component we are using for the pause and stop button: https://github.com/SmartManoj/Kevin/pull/1

File: frontend/src/components/chat/ChatInterface.tsx
Comment (SmartManoj) [reply frontend/src/components/chat/ChatInterface.tsx]: Auto mode or Auto Mode?

**Comment by xingyaoww** [comment]
My input is: "Can you build a website for me?"

Then i hit with the following errors: 
image

File: agenthub/codeact_agent/codeact_agent.py
Comment (xingyaoww) [inline agenthub/codeact_agent/codeact_agent.py:231]: I think you probably don't need to change these on the backend, since you are already dispatching user messages on the FE?

File: opendevin/core/config.py
Comment (xingyaoww) [inline opendevin/core/config.py:116]: And why do we need this? I think `is_autonoumous` can be entirely handled by the FE - no?
e.g., when the autonomous mode is enabled, whenever the agent is requesting user inputs, the FE will always send back a pre-defined user message, asking the agent to "continue". We can probably re-use the continue button logic - the "checkbox" will just work by keep clicking the "continue" button for the user.

File: agenthub/codeact_agent/codeact_agent.py
Comment (SmartManoj) [reply agenthub/codeact_agent/codeact_agent.py:231]: Not dispatched.

File: opendevin/core/config.py
Comment (SmartManoj) [reply opendevin/core/config.py]: Removed the config and added it as an attribute to Agent.  The Agent should behave differently when autonomous mode is enabled like when to detect the loop.

File: agenthub/codeact_agent/codeact_agent.py
Comment (xingyaoww) [reply agenthub/codeact_agent/codeact_agent.py:230]: Can you remove changes from the backend so that it won't affect other things (e.g., eval), and move this message to dispatched by the frontend?

Ideally, the "auto mode" should just be a way for the user to say "i want to keep clicking the [continue] button when it comes up"

File: opendevin/core/config.py
Comment (xingyaoww) [reply opendevin/core/config.py]: Why we need to change the backend for the autonomous mode?

ref: https://github.com/OpenDevin/OpenDevin/pull/2435#discussion_r1664446111

Comment (SmartManoj) [reply opendevin/core/config.py]: for different prompts?

Comment (xingyaoww) [reply opendevin/core/config.py]: what do you mean by "for different prompts"? can you elaborate?

File: agenthub/codeact_agent/codeact_agent.py
Comment (SmartManoj) [reply agenthub/codeact_agent/codeact_agent.py:230]: eval is not affected by this.

File: opendevin/core/config.py
Comment (SmartManoj) [reply opendevin/core/config.py]: for different system prompts.

Comment (xingyaoww) [reply opendevin/core/config.py]: why? I don't want us to add unnecessary additional complexity to the repo unless there are a few specific examples showing the necessity. The "continue" button on FE does not need to be modified on the backend; why does this need to modify BE?

**Comment by xingyaoww** [comment]
Close this in favor of https://github.com/OpenDevin/OpenDevin/pull/2782.

@SmartManoj  I think we *might* need the agent to be aware of this, but i don't yet see a clear reason / motivating example of doing so.

**Pull request closed without merging**
